### PR TITLE
add entry-tree-cache and gossip rfc

### PR DESCRIPTION
**Pull request by rob-solana**

### CODE REVIEW

File: rfcs/0011-db-ledger.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+Blobs will be moved to a fork-able key space the tuple of `leader slot` + `blob index` (within the slot).  This permits the skip-list structure of the Solana protocol to be stored in its entirety, without a-priori choosing which fork to follow, which Entries to persist or when to persist them.
+
+Repair requests for recent blobs are served out of RAM or recent files and out of deeper storage for less recent blobs, as implemented by the store backing DbLedger.
+
+### Functionalities of DbLedger (and hereafter "Ledger")
+
+1. Persistence: the ledger lives in the front of the nodes verification pipeline, right behind network receive and signature verification.  If the blob received is consistent with the leader schedule (i.e. was signed by the leader for the indicated slot), it is immediately stored.
+2. Repair: repair is the same as window repair above, but able to serve any blob that's been received. The ledger stores blobs with signatures, preserving the chain of origination.
+3. Forks: the ledger supports random access of blobs, so can support a validator's need to rollback and replay from a Bank checkpoint.
+4. Restart: the
```
Comment (garious) [inline rfcs/0011-db-ledger.md]: the?

**Comment by garious** [comment]
This is really nice! Thanks for the patient description of the current design and its limitations. What's still confusing to me is the jump between DbLedger to Ledger. I wonder if perhaps DbLedger could be named something like EntryTreeCache? And that a ledger would be defined as a particular path up that tree from the genesis block. That is, a ledger is still just a sequence of entries, but that we need to maintain a cache of entries, stitch them into a tree, and prune branches when lockout allows for it, and boot the trunk when it's no longer useful.

**Comment by rob-solana** [comment]
I like your new names, I "dbledger" considered harmful FTW.

File: rfcs/0011-db-ledger.md (hunk 1 above)
Comment (rob-solana) [reply rfcs/0011-db-ledger.md]: thought I'd updated this...  is kinda TBD, will finish.

**Comment by rob-solana** [comment]
@sakridge @aeyakovenko I've made you assignees because I want comments, discussion in this PR.  this is the method @garious wants us to use for book content, design updates, etc.  thanks in advance.

File: rfcs/0012-gossip.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+Nodes continuously share signed data objects among themselves in order to find each other, structure the network data plane, prevent censoring, etc.
+
+The basic mechanism is a 10Hz loop wherein every node sends a push message and/or a pull message.  Push and pull messages may elicit responses, and push messages may be forwarded on to others in the network.
+
+Gossip runs on a well-known UDP/IP port, or a port in a well-known range.  Once a network is bootstrapped, nodes advertise to each other where to find their gossip endpoint (a socket address).
+
+## Gossip Records
+
+Records shared over gossip are arbitrary, but signed and versioned (with a timestamp), and need to make sense to the node receiving them.  A strategy of "most recently recevied wins" is employed to manage what's stored in cluster info.
```
Comment (aeyakovenko) [inline rfcs/0012-gossip.md]: @rob-solana Its most recent timestamp which is signed by the creator, not most recently received.

File: rfcs/0012-gossip.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+   b. stores the message in "pushed once" (used for detecting duplicates)
+   c. retransmits the messages to its own push peers
+
+3. timeout: a push message has a time to live, beyond which the message is dropped
+
+### Push Peers, Prune Message
+
+A nodes selects its push peers at random from the set of known peers.  The node keeps this selection for a relatively long time.  When a prune message is received, the node drops the push peer that sent the prune.  Prune is an indication that there is another, faster path to that node than direct push.
+
+A new set of push peers is selected (TODO: when is that??).
```
Comment (aeyakovenko) [inline rfcs/0012-gossip.md]: rotate 1 node every `CRDS_GOSSIP_PULL_CRDS_TIMEOUT_MS/2`

Comment (rob-solana) [reply rfcs/0012-gossip.md]: sounds good.  looks like this code isn't present today...

Comment (aeyakovenko) [reply rfcs/0012-gossip.md]: @rob-solana https://github.com/solana-labs/solana/blob/master/src/cluster_info.rs#L96

Comment (rob-solana) [reply rfcs/0012-gossip.md]: yeah, but that's apparently only called once per cluster_info instantiation.  also, it's a complete reboot of push peers, right?

Comment (rob-solana) [reply rfcs/0012-gossip.md]: That's not a complete reboot of push peers?

Comment (aeyakovenko) [reply rfcs/0012-gossip.md]: https://github.com/solana-labs/solana/blob/master/src/cluster_info.rs#L571

Comment (aeyakovenko) [reply rfcs/0012-gossip.md]: @rob-solana its not, it just reboots 1.  the number is a denominator of how many to refresh.

**Pull request merged**